Enhance notify2

yccheok opened the pull request.

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,106 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     *
+     * @param tag unique identifier of the section
+     */
+    public void notifyVisibilityChangedToVisible(String tag) {
```
please change method name to `notifySectionChangedToVisible`

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,106 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+    }
+
+    /**
+     * Helper method that calls {@link #notifyItemInserted} when section changed to visible.
```
same regarding javadoc and here it should be `notifyItemRangeInserted`

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,106 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+    }
+
+    /**
+     * Helper method that calls {@link #notifyItemRangeRemoved} when section changed to invisible.
```
same regarding javadoc

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,106 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     * @param tag unique identifier of the section
+     * @param previousSectionPosition previous section position using {@link #getSectionPosition}
+     */
+    public void notifyVisibilityChangedToInvisible(String tag, int previousSectionPosition) {
```
same regarding method name

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,106 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     * Helper method that calls {@link #notifyItemRangeRemoved} when section changed to invisible.
+     *
+     * @param tag unique identifier of the section
+     * @param previousSectionPosition previous section position using {@link #getSectionPosition}
```
I would remove the "using {@link #getSectionPosition}" from here because is not clear that the user should call it before setting the position to false. Move this explanation to the description of the method 👍

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,106 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+    }
+
+    /**
+     * Helper method that calls {@link #notifyItemInserted} when section changed to visible.
```
same regarding javadoc and `notifyItemRangeInserted`

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,106 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     *
+     * @param section a visible section of this adapter
+     */
+    public void notifyVisibilityChangedToVisible(Section section) {
```
same regarding method name

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,106 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     */
+    public void notifyVisibilityChangedToVisible(Section section) {
+        if (!section.isVisible()) {
+            throw new IllegalStateException("Invalid visibility");
```
Please change message to "This section is not visible."

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,106 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+    }
+
+    /**
+     * Helper method that calls {@link #notifyItemRangeRemoved} when section changed to invisible.
```
same regarding javadoc

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,106 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     * Helper method that calls {@link #notifyItemRangeRemoved} when section changed to invisible.
+     *
+     * @param section an invisible section of this adapter
+     * @param previousSectionPosition previous section position using {@link #getSectionPosition}
```
same regarding "using .."

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,106 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     * @param section an invisible section of this adapter
+     * @param previousSectionPosition previous section position using {@link #getSectionPosition}
+     */
+    public void notifyVisibilityChangedToInvisible(Section section, int previousSectionPosition) {
```
same regarding method name

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,106 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     */
+    public void notifyVisibilityChangedToInvisible(Section section, int previousSectionPosition) {
+        if (section.isVisible()) {
+            throw new IllegalStateException("Invalid visibility");
```
same regarding message

yccheok commented:
Thanks. I had rectified. Not sure I had missed out anything...

luizgrp commented:
@yccheok are you planning to add the following methods?
- notifyHeaderRemovedFromSection 
- notifyFooterRemovedFromSection 

If so, are you planning to add to this PR or in a next one?

yccheok commented:
Opps. I think I had missed out

    notifyHeaderRemovedFromSection
    notifyFooterRemovedFromSection

I will add them in this PR "Enhance notify2"

yccheok commented:
Hi @luizgrp 

Regarding notification when footer become invisible.

Should the code be

    public void notifyFooterRemovedFromSection(int previousFooterPosition) {
        callSuperNotifyItemRemoved(previousFooterPosition);
    }

or

    public void notifyFooterRemovedFromSection(Section section) {
        int position = getSectionPosition(section) + section.getSectionItemsTotal();

        callSuperNotifyItemRemoved(position);
    }

If the code is later, do we need to concern on the API usage correctness?

(Correct usage)
    section.setHasHeader(false);
    notifyHeaderRemovedFromSection(section);
    section.setHasFooter(false);
    notifyFooterRemovedFromSection(section);

(I'm not too sure whether this will yield correct behavior. I'm concern on `getSectionItemsTotal` value returned in `notifyFooterRemovedFromSection`)
    section.setHasHeader(false);
    section.setHasFooter(false);
    notifyHeaderRemovedFromSection(section);
    notifyFooterRemovedFromSection(section);

luizgrp commented:
I'd say the latter.

Yeah, there are many ways of calling the notify methods wrongly, it's out of our control. Same for the notify methods of the native RecyclerView Adapter. 
The helper methods are provided to help with the calculation of the positions but it's up to the developers to call them appropriately.

yccheok commented:
Great. I will then go ahead and change accordingly.

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,122 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     * @param tag unique identifier of the section
+     */
+    public void notifyHeaderInsertedInSection(String tag) {
+        int headerPosition = getHeaderPositionInAdapter(tag);
```
shouldn't we follow the pattern of calling `getValidSectionOrThrowException` then  the corresponded `notifyHeaderInsertedInSection` method with Section as parameter?

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,122 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     * @param section a visible section of this adapter
+     */
+    public void notifyHeaderInsertedInSection(Section section) {
+        int headerPosition = getHeaderPositionInAdapter(section);
```
same here

yccheok replied:
I think you mean `public void notifyFooterInsertedInSection(String tag)` right?

luizgrp replied:
yes, sorry!

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,122 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     * @param tag unique identifier of the section
+     */
+    public void notifySectionChangedToVisible(String tag) {
+        Section section = getSection(tag);
```
getValidSectionOrThrowException

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,122 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     */
+    public void notifySectionChangedToVisible(Section section) {
+        if (!section.isVisible()) {
+            throw new IllegalStateException("This section is not visible.");
```
missing unit test for this exception

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,122 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     */
+    public void notifySectionChangedToInvisible(Section section, int previousSectionPosition) {
+        if (section.isVisible()) {
+            throw new IllegalStateException("This section is not visible.");
```
same

luizgrp commented on library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java:
```diff
@@ -841,6 +841,160 @@ public void notifyItemMovedInSectionUsingSection_withAdapterWithManySections_cal
+    }
+
+    @Test
+    public void notifyHeaderInsertedInSectionUsingSection_withAdapterWithManySections_callsSuperNotifyItemInserted() {
```
redundant test, previous one is also testing the corresponded method that has section as parameter

luizgrp commented on library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java:
```diff
@@ -841,6 +841,160 @@ public void notifyItemMovedInSectionUsingSection_withAdapterWithManySections_cal
+    }
+
+    @Test
+    public void notifyFooterInsertedInSectionUsingSection_withAdapterWithManySections_callsSuperNotifyItemInserted() {
```
same - redundant

luizgrp commented on library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java:
```diff
@@ -841,6 +841,160 @@ public void notifyItemMovedInSectionUsingSection_withAdapterWithManySections_cal
+    }
+
+    @Test
+    public void notifySectionChangedToVisibleUsingSection_withAdapterWithManySections_callsSuperNotifyItemRangeInserted() {
```
same - redundant

luizgrp commented on library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java:
```diff
@@ -841,6 +841,160 @@ public void notifyItemMovedInSectionUsingSection_withAdapterWithManySections_cal
+    }
+
+    @Test
+    public void notifySectionChangedToInvisibleUsingSection_withAdapterWithManySections_callsSuperNotifyItemRangeInserted() {
```
same - redundant

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,122 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     *
+     * @param section a visible section of this adapter
+     */
+    public void notifySectionChangedToVisible(Section section) {
```
can you please move this method to under the corresponded  `notifySectionChangedToVisible(tag)` ?

luizgrp commented on library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java:
```diff
@@ -793,6 +793,122 @@ void callSuperNotifyItemMoved(int fromPosition, int toPosition) {
+     * @param section an invisible section of this adapter
+     * @param previousSectionPosition previous section position
+     */
+    public void notifySectionChangedToInvisible(Section section, int previousSectionPosition) {
```
and as result this one should be under the corresponded `notifySectionChangedToInvisible(tag)`

yccheok commented:
@luizgrp 

Added the missing functions and UT

Note that, in https://github.com/luizgrp/SectionedRecyclerViewAdapter/pull/65/commits/23b085616ce4476e39c7c2ebe968c0998c671c09 , I changed `getSection` to `getValidSectionOrThrowException`

luizgrp commented on library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java:
```diff
@@ -841,6 +841,236 @@ public void notifyItemMovedInSectionUsingSection_withAdapterWithManySections_cal
+    }
+
+    @Test
+    public void notifyHeaderRemovedFromSectionUsingSection_withAdapterWithManySections_callsSuperNotifyItemRemoved() {
```
same - redundant

luizgrp commented on library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java:
```diff
@@ -841,6 +841,236 @@ public void notifyItemMovedInSectionUsingSection_withAdapterWithManySections_cal
+    }
+
+    @Test
+    public void notifyFooterRemovedFromSectionUsingSection_withAdapterWithManySections_callsSuperNotifyItemRemoved() {
```
same - redundant

yccheok commented:
Changed according to reviews :)

luizgrp commented on library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java:
```diff
@@ -841,6 +841,152 @@ public void notifyItemMovedInSectionUsingSection_withAdapterWithManySections_cal
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void notifySectionChangedToVisibleUsingTag_withAdapterWithManySections_throwsException() {
```
notifySectionChangedToVisibleUsingTag_**withInvisibleSection**_throwsException

luizgrp commented on library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java:
```diff
@@ -841,6 +841,152 @@ public void notifyItemMovedInSectionUsingSection_withAdapterWithManySections_cal
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void notifySectionChangedToInvisibleUsingTag_withAdapterWithManySections_throwsException() {
```
notifySectionChangedToInvisibleUsingTag_**withVisibleSection**_throwsException

luizgrp commented on library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java:
```diff
@@ -841,6 +841,152 @@ public void notifyItemMovedInSectionUsingSection_withAdapterWithManySections_cal
+        SectionedRecyclerViewAdapter spySectionedRecyclerViewAdapter = spy(SectionedRecyclerViewAdapter.class);
+        doNothing().when(spySectionedRecyclerViewAdapter).callSuperNotifyItemRangeInserted(anyInt(), anyInt());
+
+        spySectionedRecyclerViewAdapter.addSection(new StatelessSectionStub(ITEMS_QTY));
```
not needed for this test as we are not testing any position

luizgrp commented on library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java:
```diff
@@ -841,6 +841,152 @@ public void notifyItemMovedInSectionUsingSection_withAdapterWithManySections_cal
+        SectionedRecyclerViewAdapter spySectionedRecyclerViewAdapter = spy(SectionedRecyclerViewAdapter.class);
+        doNothing().when(spySectionedRecyclerViewAdapter).callSuperNotifyItemRangeRemoved(anyInt(), anyInt());
+
+        spySectionedRecyclerViewAdapter.addSection(new StatelessSectionStub(ITEMS_QTY));
```
same - not needed

luizgrp commented:
this is good to go after addressing comments from last review 👍

yccheok commented:
Changed according to review.

luizgrp commented:
@yccheok this branch got conflicts after merging the other one

yccheok commented:
I created another pull request to solve the conflict.

https://github.com/luizgrp/SectionedRecyclerViewAdapter/pull/69
